propagation/w3c.go

```diff
@@ -0,0 +1,165 @@
+package propagation
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel/api/kv"
+	"go.opentelemetry.io/otel/api/trace"
+	"google.golang.org/grpc/codes"
+)
+
+// MarshalHoneycombTraceContext uses the information in prop to create trace context headers
+// that conform to the W3C Trace Context specification. The header values are set in headers,
+// which is an HTTPSupplier, an interface to which http.Header is an implementation. The headers
+// are also returned as a map[string]string.
+func MarshalW3CTraceContext(ctx context.Context, prop *PropagationContext) (context.Context, map[string]string) {
```

Comment:
Can you say something about why the function signature for marshaling w3c is different from marshaling honeycomb or aws headers? It feels like if we have several functions `MarshalFooTraceContext` they should all have the same signature to make clear that they are doing the same thing. If they are not doing the same thing, is there an abstraction layer here that's leaking? 
This may be resolved with some example code once we get further in the process...

Reply:
They're different because W3C Trace Context can produce multiple headers, whereas both the Amazon and Honeycomb marshal functions produce only one. A possible solution is to make the Amazon and Honeycomb marshal functions also return a `map[string]string`. Additionally, the W3C Trace Context functions take and return a Context because those are used by the OTel SDK. Again, could just add the context to the Amazon and Honeycomb marshal / unmarshal functions even if it's ignored.

Reply:
Done. Also wrote up some alternative options in an internal quip document. Happy to discuss.

Reply:
On second thought, maybe it's a _good thing_ that the function signature for one is different. These are all expected to be called from within the hook (that will have access to the full request object, including its context). That the specifics of each protocol might need different things from the request seems reasonable - and provides a good example for someone writing some other serialization code later that does something else funky.
When I was asking about the abstraction layer above - the marshal / unmarshal functions are below the consistent portion of the abstraction (which is the callback hook). No leaking here! (maybe.)

Reply:
This is the conclusion I've come to. I think the hooks need to be consistent, the marshal / unmarshal functions do not. They should be close in behavior but don't need to be consistent in the interface.